### PR TITLE
PYTHON-4667 Handle  $clusterTime from error responses in client Bulk Write

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: pymongo/asynchronous/client_bulk.py [hunk 1]
```diff
@@ -312,6 +312,9 @@ async def write_command(
                 bwc._fail(request_id, failure, duration)
             # Top-level error will be embedded in ClientBulkWriteException.
             reply = {"error": exc}
+            # Propagate $clusterTime to the caller.
+            if "$clusterTime" in cmd:
+                reply["$clusterTime"] = cmd["$clusterTime"]
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/client_bulk.py:317]: $clusterTime needs to be parsed from the (potential) server error response, not the command. We also need to do the same for operationTime.

This is the method that handles it, it probably makes more sense to call this method at a lower level if we can (perhaps in write_command):
```
client._process_response(command_response, session)
```

Or another alternative.

Comment (blink1073) [reply pymongo/asynchronous/client_bulk.py:317]: Makes sense, updated.

File: pymongo/asynchronous/bulk.py [hunk 2]
```diff
@@ -308,6 +309,11 @@ async def write_command(
 
             if bwc.publish:
                 bwc._fail(request_id, failure, duration)
+            # Process the response from the server.
+            if isinstance(exc, OperationFailure):
+                await client._process_response(exc.details, bwc.session)  # type: ignore[arg-type]
+            else:
+                await client._process_response({}, bwc.session)  # type: ignore[arg-type]
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/bulk.py:316]: We don't need an `else` here.

File: pymongo/asynchronous/bulk.py [hunk 3]
```diff
@@ -308,6 +309,11 @@ async def write_command(
 
             if bwc.publish:
                 bwc._fail(request_id, failure, duration)
+            # Process the response from the server.
+            if isinstance(exc, OperationFailure):
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/bulk.py:313]: `isinstance(exc, (NotPrimaryError, OperationFailure))`

File: pymongo/synchronous/bulk.py [hunk 4]
```diff
@@ -308,6 +309,11 @@ def write_command(
 
             if bwc.publish:
                 bwc._fail(request_id, failure, duration)
+            # Process the response from the server.
+            if isinstance(exc, OperationFailure):
```
Comment (ShaneHarvey) [inline pymongo/synchronous/bulk.py:313]: Same comments here.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Could you add a test to `test/mockupdb/test_cluster_time.py`? Something like:
```python
    def test_bulk_error(self):
        def callback(client: MongoClient[dict]) -> None:
            client.db.collection.bulk_write(
                [InsertOne({}), InsertOne({})]
            )

        self.cluster_time_conversation(
            callback,
            [{"ok": 0, "errmsg": "mock error"}],
        )
```

File: pymongo/asynchronous/bulk.py (hunk 2 above)
Comment (blink1073) [reply pymongo/asynchronous/bulk.py:316]: Removed

File: pymongo/asynchronous/bulk.py (hunk 3 above)
Comment (blink1073) [reply pymongo/asynchronous/bulk.py]: Done

File: pymongo/synchronous/bulk.py (hunk 4 above)
Comment (blink1073) [reply pymongo/synchronous/bulk.py]: Automatically done.  :smile:

**Comment by blink1073** [comment]
> Could you add a test

Done

File: test/mockupdb/test_cluster_time.py [hunk 5]
```diff
@@ -166,6 +167,17 @@ def test_monitor(self):
...
+    def test_bulk_error(self):
+        def callback(client: MongoClient[dict]) -> None:
+            client.db.collection.bulk_write([InsertOne({}), InsertOne({})])
+
+        with self.assertRaises(OperationFailure) as context:
+            self.cluster_time_conversation(
+                callback,
+                [{"ok": 0, "errmsg": "mock error"}],
+            )
+        self.assertIn("$clusterTime", str(context.exception))
```
Comment (ShaneHarvey) [inline test/mockupdb/test_cluster_time.py:179]: This test is skipping the clusterTime asserts in cluster_time_conversation. We need to do:
```
        def callback(client: MongoClient[dict]) -> None:
            with self.assertRaises(OperationFailure):
                client.db.collection.bulk_write([InsertOne({}), InsertOne({})])

        self.cluster_time_conversation(
            callback,
            [{"ok": 0, "errmsg": "mock error"}],
        )
```

Comment (ShaneHarvey) [reply test/mockupdb/test_cluster_time.py:179]: Could you also add a similar test for client.bulk_write?

Comment (blink1073) [reply test/mockupdb/test_cluster_time.py]: Done and done

**Review by ShaneHarvey — APPROVED** [review]
LGTM once the tests pass!

**Pull request merged**